### PR TITLE
Fixes update service endpoints in place (resolves recent vmi networking test failures)

### PR DESCRIPTION
**Pull request by davidvossel**
We've recently noticed some odd behavior in our Service endpoints. This behavior was exasperated further by a change to our operator that caused resources to be acted upon more frequently.

The issue is that ClusterIP service endpoints are not Patched in place. Instead they are deleted and re-created. This causes downtime to our API and can cause transient issues in our functional testsuite. 

This issue appeared releated to the recent CustomizePatches PR #3612 but was in reality completely unrelated. That PR did however introduce a naming collision with our api. KubeVirt v1.Patch and Kubernetes v1.Patch both exist. I fixed this by simply renaming our struct so it doesn't collide. 

**Release note**:

```release-note
Fixes issue with Serivce endpoints not being updated properly in place during KubeVirt updates. 
```

### CODE REVIEW

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 1]
```diff
@@ -1172,11 +1172,11 @@ type KubeVirtSpec struct {
 // +k8s:openapi-gen=true
 type CustomizeComponents struct {
 	// +listType=atomic
-	Patches []Patch `json:"patches,omitempty"`
+	Patches []CustomizeComponentsPatch `json:"patches,omitempty"`
 }
 
 // +k8s:openapi-gen=true
-type Patch struct {
+type CustomizeComponentsPatch struct {
```
Comment (xpivarc) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:1179]: I am all for this. But we should consider to improve our generating of openapi so this problem doesn't come up in the future.

File: pkg/virt-operator/install-strategy/create.go [hunk 2]
```diff
@@ -1415,10 +1419,18 @@ func generateServicePatch(kv *v1.KubeVirt,
 		deleteAndReplace = true
 	} else if cachedService.Spec.ClusterIP != service.Spec.ClusterIP {
 
-		// clusterIP is not mutable. A ClusterIP == "" will mean one is dynamically assigned.
-		// It is okay if the cached service has a ClusterIP and the target one is empty. We just
-		// ensure the cached ClusterIP is carried over in the Patch.
-		deleteAndReplace = true
+		if service.Spec.ClusterIP == "" {
```
Comment (xpivarc) [inline pkg/virt-operator/install-strategy/create.go:1422]: Why do we rely on ClusterIP?

**Comment by jean-edouard** [comment]
/approve
Good catch(es)! Looks like there are more instances of `Patch` to fix (seen in travis-ci: `pkg/virt-operator/install-strategy/patches.go:21:12: undefined: "kubevirt.io/client-go/api/v1".Patch`).
Approving ahead of that getting addressed (as well as the other comments above) , since the general idea makes sense and we need this asap.

**Comment by davidvossel** [comment]
> Good catch(es)! Looks like there are more instances of Patch to fix (seen in travis-ci: pkg/virt-operator/install-strategy/patches.go:21:12: undefined: "kubevirt.io/client-go/api/v1".Patch).

yup, fixed. I added some unit tests as well for the ClusterIP logic.

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 1 above)
Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1179]: yes. either improve our openapi generation logic, or create some sort of test case that caches when we collide with a Kubernetes api field.

File: pkg/virt-operator/install-strategy/create.go (hunk 2 above)
Comment (davidvossel) [reply pkg/virt-operator/install-strategy/create.go:1422]: Service resources are a bit strange.  If you set the ClusterIP == "", then the k8s backend automatically assigns an IP address. The weird thing here for us is that the k8s backend actually mutates the service's spec with that new ClusterIP field.

So, the install strategy's desired service has `ClusterIP == ""` and the cached service has `ClusterIP == <some dynamic IP>`

Typically for our other resources we'd simply stomp on any changes that differ in the spec from our desired spec, versus what we see cached. However for a ClusterIP service, we actually need to preserve the cached ClusterIP field and merge it with the install strategies desired service spec.

Basically, cluster ip is something that we set to "" and magically turns into an IP later on. This inconsistency between our desired spec and what it gets mutated into requires some special handling.

Comment (xpivarc) [reply pkg/virt-operator/install-strategy/create.go:1422]: Thanks, that makes absolute sense :+1:

**Review by xpivarc — COMMENTED** [review]
There are some leftovers, otherwise
\lgtm

File: api/openapi-spec/swagger.json [hunk 3]
```diff
@@ -6632,12 +6632,29 @@
...
       "type": "array",
       "items": {
-       "$ref": "#/definitions/v1.Patch"
+       "$ref": "#/definitions/v1.CustomizeComponentsPatch"
       },
       "x-kubernetes-list-type": "atomic"
      }
     }
    },
+   "v1.CustomizeComponentsPatch": {
```
Comment (oshoval) [inline api/openapi-spec/swagger.json:6641]: Please revert this file, unless you did it by purpose of this PR.
Reason is there is a known issue of flakiness generating swagger.json

File: pkg/virt-operator/install-strategy/create_test.go [hunk 4]
```diff
@@ -278,6 +279,83 @@ var _ = Describe("Create", func() {
...
+				},
+				Spec: v1.KubeVirtSpec{
+					ImageRegistry: "someregistery",
+					ImageTag:      "v1",
+				},
+			}
+
+			cachedService := &corev1.Service{}
+			cachedService.Spec.Type = corev1.ServiceTypeClusterIP
+			cachedService.Spec.ClusterIP = "10.10.10.10"
```
Comment (oshoval) [inline pkg/virt-operator/install-strategy/create_test.go:298]: do we need to add ipv6 test here ? or its not important ?

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 1 above)
Comment (dhiller) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1179]: I'd prefer the latter.

File: pkg/virt-operator/install-strategy/create_test.go (hunk 4 above)
Comment (davidvossel) [reply pkg/virt-operator/install-strategy/create_test.go:298]: not important. i could set this to anything that's != "" for the purposes here.

File: api/openapi-spec/swagger.json (hunk 3 above)
Comment (davidvossel) [reply api/openapi-spec/swagger.json:6641]: This pr is meant to address all the current dev issues we're encountering. This fixes the flaky swagger generation logic... at least for now. A better fix is welcome in the future, this is a simply way to get unblocked though that doesn't impact anything reallly.
